numpyro/primitives.py

```diff
@@ -147,6 +147,31 @@ def param(name, init_value=None, **kwargs):
     return msg['value']
 
 
+def deterministic(name, value):
+    """
+    Used to designate deterministic sites in the model. Note that most effect
+    handlers will not operate on deterministic sites (except
+    :function:`~numpyro.handlers.trace`), so deterministic sites should be
+    side-effect free. The use case for deterministic nodes is to record any
+    values in the model execution trace.
+
+    :param str name: name of the deterministic site.
+    :param numpy.ndarray value: deterministic value to record in the trace.
```

Comment:
It would be nice if this can support a `callable` to lazily get those values (similar to `pyro.param(..., lambda: ...)`).

Reply:
Could you elaborate on this? I am not sure about why this needs to be lazy.

Reply:
There are two usage cases of `deterministic`:
+ record a value
+ record a transformed value

This PR supports the first usage case. But it will incur unnecessary overhead while running inference for the second usage case (unless users use a flag to disable the computation for inference). Nevermind though, now I think it is better to use a flag for those lazy values. It is always better to be explicit than introducing a library-knowledge pattern (especially when users ask: what is the output of those deterministic sites if its value is lazily evaluated). :)